src/visitors/TypedVisitor.ts

```diff
@@ -49,18 +50,75 @@ export interface TypedOperationDefinitionNode extends OperationDefinitionNode {
   typed: TypedField;
 }
 
+const KEYWORDS = [
+  "await",
+  "break",
+  "case",
+  "catch",
+  "class",
+  "const",
+  "continue",
+  "debugger",
+  "default",
+  "delete",
+  "do",
+  "else",
+  "enum",
+  "export",
+  "extends",
+  "false",
+  "finally",
+  "for",
+  "function",
+  "if",
+  "implements",
+  "import",
+  "in",
+  "instanceof",
+  "interface",
+  "let",
+  "new",
+  "null",
+  "package",
+  "private",
+  "protected",
+  "public",
+  "return",
+  "super",
+  "switch",
+  "static",
+  "this",
+  "throw",
+  "try",
+  "True",
+  "typeof",
+  "var",
+  "void",
+  "while",
+  "with",
+  "yield",
+];
+
+const getSafeName = (name: string) =>
+  KEYWORDS.includes(name) ? `${name}: __safe_${name}__` : name;
+
 const createTypedField = (
   field: Partial<TypedField> & { name: string }
-): TypedField => ({
-  inputFields: [],
-  outputFields: [],
-  isNonNull: false,
-  isList: false,
-  isFragment: false,
-  isTypename: false,
-  typenames: [],
-  ...field,
-});
+): TypedField => {
+  const safeName = getSafeName(field.name);
```

Comment:
That makes sense. I think I understand better now what `TypedField` is for, too. I've changed the implementation to match your suggestion.

Reply:
I think `fieldNames` is no longer the best name for this array, since now _it_ sometimes contains JS constructs, but I'm struggling to think of a better label. How do you feel about `safelyReferencedFieldNames`?

We could optionally do the safe-naming in the two places we iterate over `fieldNames`, which seems more _correct_, but I gave it a try and it doesn't feel great to me.

Reply:
True, when proposing the change I was also not sure how to call this variable :) 
I think one reason why it is difficult to name it, is that the resulting expression is used for two different js language constructs: once for a destructuring assignment and once for initializing an object. Incidentally, our expression fits for both.

I have an idea to make this difference explicit in the code which could also make naming simpler. But please consider this just an idea, I think also your current solution is great, I leave this decision up to you.

Since we construct expressions, and not names, I would call them accordingly. Also I would split into two variables, one for each expression type, even if that leads to some repetition (lookup of reserved names and formatting).

```
const destructuringFields = fields.filter((field) => !field.isFragment)
const destructuringExpressions = destructuringFields.map(...)
const objectInitializationExpressions = destructuringFields.map(...)
```

Reply:
Yes! That's more like what I was groping towards. Expressions, not names.

Reply:
Here's my take on that https://github.com/rubengrill/apollo-typed-documents/pull/57/commits/0f1d1a301a0746419660359838fe0d08c79a7aa7

I put `safeNames` back just the way it was before my changes and added a `getSafeAlias` function to dedupe and help explain why the variable is assigned to a `__weird_name__`.

Reply:
@ccblaisdell Awesome, `getSafeAlias` makes it really clear 👍